indexing-service/src/main/java/org/apache/druid/indexing/common/task/KillUnusedSegmentsByIdTask.java

```diff
@@ -0,0 +1,199 @@
+
+  @JsonCreator
+  public KillUnusedSegmentsByIdTask(
+      @JsonProperty("id") String id,
+      @JsonProperty("dataSource") String dataSource,
+      @JsonProperty("context") Map<String, Object> context,
+      @JsonProperty("segmentIds") List<String> segmentIds
+  )
+  {
+    this(
+        getOrMakeId(id, TYPE, dataSource, null),
+        dataSource,
+        segmentIds,
+        context
+    );
+  }
+
+  protected KillUnusedSegmentsByIdTask(
+      String id,
+      String dataSource,
+      List<String> segmentIds,
+      Map<String, Object> context
+  )
+  {
+    this(id, id, new TaskResource(id, 1), dataSource, segmentIds, context);
+  }
+
+  protected KillUnusedSegmentsByIdTask(
+      String id,
+      String groupId,
+      TaskResource taskResource,
+      String dataSource,
+      List<String> segmentIds,
+      Map<String, Object> context
+  )
+  {
+    super(id, groupId, taskResource, dataSource, context);
+    this.segmentIds = Preconditions.checkNotNull(segmentIds, "segmentIds");
+  }
+
+  @JsonProperty
+  public List<String> getSegmentIds()
+  {
+    return segmentIds;
+  }
+
+  @Override
+  public String getType()
+  {
+    return TYPE;
+  }
+
+  @Nonnull
+  @JsonIgnore
+  @Override
+  public Set<ResourceAction> getInputSourceResources()
+  {
+    return ImmutableSet.of();
+  }
+
+  @Override
+  public TaskStatus runTask(TaskToolbox toolbox) throws Exception
+  {
+    final NavigableMap<DateTime, List<TaskLock>> taskLockMap = getTaskLockMap(toolbox.getTaskActionClient());
+
+    // List unused segments
+    List<DataSegment> unusedSegments;
+    synchronized (this) {
+      unusedSegments = ImmutableList.copyOf(unusedSegmentsTokill);
+    }
+
+    if (!TaskLocks.isLockCoversSegments(taskLockMap, unusedSegments)) {
+      throw new ISE(
+          "Locks[%s] for task[%s] can't cover segments[%s]",
+          taskLockMap.values().stream().flatMap(List::stream).collect(Collectors.toList()),
+          getId(),
+          unusedSegments
+      );
+    }
+
+    // Kill segments
+    toolbox.getTaskActionClient().submit(new SegmentNukeAction(new HashSet<>(unusedSegments)));
+    for (DataSegment segment : unusedSegments) {
+      toolbox.getDataSegmentKiller().kill(segment);
+    }
+
+    return TaskStatus.success(getId());
+  }
+
+  @Override
+  public boolean isReady(TaskActionClient taskActionClient) throws Exception
+  {
+    List<DataSegment> unusedSegments =
+        taskActionClient.submit(new RetrieveUnusedSegmentsByIdAction(getDataSource(), getSegmentIds()));
+    for (DataSegment unusedSegment : unusedSegments) {
+      final TaskLock lock = taskActionClient.submit(
+          new TimeChunkLockTryAcquireAction(
+              TaskLockType.EXCLUSIVE,
+              unusedSegment.getInterval()
+          )
+      );
+      if (lock == null) {
+        return false;
+      }
+      if (lock.isRevoked()) {
+        throw new ISE(StringUtils.format("Lock for interval [%s] was revoked.", unusedSegment.getInterval()));
```

Comment:
if it can't get a lock for a certain interval, can it just skip that segment instead of failing the whole task?

Reply:
in the context of this being issued by the coordinator auto-kill, IMO it feels like it would be better to let the task partially succeed and get rid of what it can, instead of requiring that it kill everything or nothing

Reply:
or be blocked on waiting for a lock to be released